From 28467b83ff84f85d8f33a0468eed8de923b4d0e9 Mon Sep 17 00:00:00 2001
From: Rany Hany <rany_hany@riseup.net>
Date: Wed, 19 Jun 2024 00:56:24 +0300
Subject: [PATCH] nl80211: handle per_sta_vif case properly

When per_sta_vif is enabled in hostapd BSS config, we need to iterate
through every interface in /sys/class/net which matches
`${interface_requested}.${number}`. Otherwise data requested from iwinfo
will be incomplete/nonexistant. This is already handled for WDS so there is
precedence for a change like this.

Also refactor logic on whether to issue a NL80211_CMD_GET_STATION request
on the interface into a seperate function.

Closes https://github.com/openwrt/openwrt/issues/14339

Signed-off-by: Rany Hany <rany_hany@riseup.net>
---
 iwinfo_nl80211.c | 42 ++++++++++++++++++++++++++++++++++++------
 1 file changed, 36 insertions(+), 6 deletions(-)

diff --git a/iwinfo_nl80211.c b/iwinfo_nl80211.c
index 2200249..9469af3 100644
--- a/iwinfo_nl80211.c
+++ b/iwinfo_nl80211.c
@@ -153,6 +153,38 @@ static int nl80211_readstr(const char *path, char *buffer, int length)
 	return rv;
 }
 
+static bool nl80211_is_numeric(const char *str)
+{
+	return *str && !str[strspn(str, "0123456789")];
+}
+
+static bool nl80211_strstarts(const char *str, const char *prefix, size_t len)
+{
+	return strncmp(str, prefix, len) == 0;
+}
+
+static bool nl80211_should_get_station(const char *cur_name, const char *req_name,
+				       size_t req_name_len)
+{
+	if (!nl80211_strstarts(cur_name, req_name, req_name_len))
+		return false;
+
+	if (!cur_name[req_name_len])
+		return true;
+
+	if (cur_name[req_name_len] == '.')
+	{
+		if (nl80211_strstarts(&cur_name[req_name_len + 1], "sta", 3) &&
+		    nl80211_is_numeric(&cur_name[req_name_len + 4]))
+			return true;
+
+		if (nl80211_is_numeric(&cur_name[req_name_len + 1]))
+			return true;
+	}
+
+	return false;
+}
+
 static int nl80211_get_band(int nla_type)
 {
 	switch (nla_type)
@@ -1608,6 +1640,7 @@ static void nl80211_fill_signal(const char *ifname, struct nl80211_rssi_rate *r)
 {
 	DIR *d;
 	struct dirent *de;
+	size_t ifname_len = strlen(ifname);
 
 	memset(r, 0, sizeof(*r));
 
@@ -1615,9 +1648,7 @@ static void nl80211_fill_signal(const char *ifname, struct nl80211_rssi_rate *r)
 	{
 		while ((de = readdir(d)) != NULL)
 		{
-			if (!strncmp(de->d_name, ifname, strlen(ifname)) &&
-			    (!de->d_name[strlen(ifname)] ||
-			     !strncmp(&de->d_name[strlen(ifname)], ".sta", 4)))
+			if (nl80211_should_get_station(de->d_name, ifname, ifname_len))
 			{
 				nl80211_request(de->d_name, NL80211_CMD_GET_STATION,
 				                NLM_F_DUMP, nl80211_fill_signal_cb, r);
@@ -2384,14 +2415,13 @@ static int nl80211_get_assoclist(const char *ifname, char *buf, int *len)
 	struct dirent *de;
 	struct nl80211_array_buf arr = { .buf = buf, .count = 0 };
 	struct iwinfo_assoclist_entry *e;
+	size_t ifname_len = strlen(ifname);
 
 	if ((d = opendir("/sys/class/net")) != NULL)
 	{
 		while ((de = readdir(d)) != NULL)
 		{
-			if (!strncmp(de->d_name, ifname, strlen(ifname)) &&
-			    (!de->d_name[strlen(ifname)] ||
-			     !strncmp(&de->d_name[strlen(ifname)], ".sta", 4)))
+			if (nl80211_should_get_station(de->d_name, ifname, ifname_len))
 			{
 				nl80211_request(de->d_name, NL80211_CMD_GET_STATION,
 				                NLM_F_DUMP, nl80211_get_assoclist_cb, &arr);
-- 
2.45.2

